rust/src/asn1/mod.rs

```diff
@@ -0,0 +1,447 @@
+    }
+
+    let slice = build_slice!(input, input_len as usize);
+
+    let ad = unsafe { &*ad_ptr };
+
+    let res = asn1_decode(slice, buffer_offset, ad);
+
+    match res {
+        Ok(asn1) => Box::into_raw(Box::new(asn1)),
+        Err(_e) => std::ptr::null_mut(),
+    }
+}
+
+/// Free a Asn1 object allocated by Rust
+///
+/// # Safety
+///
+/// ptr must be a valid object obtained using `rs_asn1_decode`
+#[no_mangle]
+pub unsafe extern "C" fn rs_asn1_free(ptr: *mut Asn1) {
+    if ptr.is_null() {
+        return;
+    }
+    drop(Box::from_raw(ptr));
+}
+
+/// This function implements the detection of the following options:
+///   - oversize_length
+///   - bitstring_overflow
+///   - double_overflow
+///
+/// # Safety
+///
+/// ptr must be a valid object obtained using `rs_asn1_decode`
+/// ad_ptr must be a valid object obtained using `rs_detect_asn1_parse`
+///
+/// Returns 1 if any of the options match, 0 if not
+#[no_mangle]
+pub unsafe extern "C" fn rs_asn1_checks(
+    ptr: *const Asn1,
+    ad_ptr: *const DetectAsn1Data,
+) -> u8 {
+    if ptr.is_null() || ad_ptr.is_null() {
+        return 0;
+    }
+
+    let asn1 = &*ptr;
+    let ad = &*ad_ptr;
+
+    match asn1.check(ad) {
+        Ok(Some(_check)) => 1,
+        Ok(None) => 0,
+        Err(e) => {
+            SCLogError!("error during asn1 checks: {}", e.to_string());
+            0
+        }
+    }
+}
+
+impl From<std::num::TryFromIntError> for Asn1DecodeError {
+    fn from(_e: std::num::TryFromIntError) -> Asn1DecodeError {
+        Asn1DecodeError::InvalidKeywordParameter
+    }
+}
+
+impl From<nom::Err<der_parser::error::BerError>> for Asn1DecodeError {
+    fn from(e: nom::Err<der_parser::error::BerError>) -> Asn1DecodeError {
+        match e {
+            nom::Err::Incomplete(_) => Asn1DecodeError::InvalidLength,
+            nom::Err::Error(e) | nom::Err::Failure(e) => match e {
+                BerError::BerTypeError => Asn1DecodeError::BerTypeError,
+                BerError::BerValueError => Asn1DecodeError::BerValueError,
+                BerError::InvalidTag => Asn1DecodeError::InvalidTag,
+                BerError::InvalidClass => Asn1DecodeError::InvalidClass,
+                BerError::InvalidLength => Asn1DecodeError::InvalidLength,
+                BerError::ConstructExpected => Asn1DecodeError::ConstructExpected,
+                BerError::ConstructUnexpected => Asn1DecodeError::ConstructUnexpected,
+                BerError::IntegerTooLarge => Asn1DecodeError::IntegerTooLarge,
+                BerError::BerMaxDepth => Asn1DecodeError::BerMaxDepth,
+                BerError::ObjectTooShort => Asn1DecodeError::ObjectTooShort,
+                BerError::DerConstraintFailed => Asn1DecodeError::DerConstraintFailed,
+                BerError::UnknownTag => Asn1DecodeError::UnknownTag,
+                BerError::Unsupported => Asn1DecodeError::Unsupported,
+                _ => Asn1DecodeError::InvalidStructure,
+            },
+        }
+    }
+}
```

Comment:
See top-file remark, I think this can be simplified by keeping the `BerError` wrapped in a variant item